File: modeling-evolution/ch5/scope.R
fun1 = function(x, a, b) { a*x + b }

fun2 = function(x) { a*x + b }

print('calling function defined earlier with all arguments passed')
for(i in 1:2) {
    for(j in 1:3) {
        a = j
        b = 2*a
        print( fun1(i, a, b) )
    }
}    

print('calling function defined earlier with unbound a and b')
for(i in 1:2) {
    for(j in 1:3) {
        a = j
        b = 2*a
        print( fun2(i) )
    }
}    

print('calling function defined locally')
for(i in 1:2) {
    for(j in 1:3) {
        c = j
        d = 2*c
        fun = function(x) { c*x + d }
        print( fun(i) )
    }
}    

print('call functions created with a and b in context; functions are stored for later call')
funs = list()
for(i in 1:3) {
    e = i
    f = 2*e
    funs[[i]] = function(x) { e*x + f }
}

print(c('during call the values of e and f are: ', e, f))
for(i in 1:2) {
    for(j in 1:3) {
        r = funs[[j]](i)
        print(r)
    }
}    

File: modeling-evolution/ch5/scope.py
fun1 = lambda x, a, b : a*x + b

fun2 = lambda x : a*x + b

def fun3(x):
  return a*x + b

print('all arguments passed')
for i in range(1,3):
  for j in range(1,4):
    a = j
    b = 2 * a
    print( fun1(i, a, b) )

print('call earlier defined function that have a, b unbound')
for i in range(1,3):
  for j in range(1,4):
    a = j
    b = 2*a
    print( fun2(i) )

print('call earlier defined function declared with def ')
for i in range(1,3):
  for j in range(1,4):
    a = j
    b = 2*a
    r = fun3(i)
    print(r)

print('stored lambdas')
funs = []
for i in range(1,4):
  a = i
  b = 2*a
  funs.append( lambda x: a * x + b )

print("since dynamic scope, the functions all use current values of a and b")
print("a=", a, "b=", b)
for i in range(1,3):
  for f in funs:
    r = f(i)
    print(r)


